### ISSUE TITLE
ControlValueAccessor with FormArray in Angular

### ISSUE THREAD

**Issue by junaidbinsarfraz** [angular/angular#31958]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/forms

### Description
 I have a child component which deals with the `FormArray` of input controls. I want to have a formcontrol over the child component.

I want to have a custom-child component say name-list. The parent `FormControl` will send `FormArray` to name-list. Then in `NameListComponent `will use `FormArray` and manipulate according to need just like `FormControl `do using `ControlValueAccessor`.

I have searched for similar functionality, I just come across [this question without legitimate answer](https://stackoverflow.com/questions/40439660/controlvalueaccessor-with-formarray-in-angular-2)

**Example**

```
<h1>Child</h1>
<div formArrayName="names" *ngFor="let name of names.controls; let i = index;">
 <div [formGroupName]="i" >
  <input formControlName="firstName">
  <input formControlName="lastName">
 </div>
</div>
```

Intention is to bind parent form with the array of input control in the child component. Also child form will have validation checks.

### Describe the solution you'd like
 Need a similar `interface ControlValueAccessor` like `ArrayAccessor

### Describe alternatives you've considered

 I have tried to send `formArrayName` to custom component, but it seems  `writeValue(obj: any)` never called for `formArrayName`, it only works for `formControlName`.

**Comment by thefliik** [comment]
My understanding of this issue is that you want to create a custom component to encapsulate a piece of a form. In this case, the form piece you wish to encapsulate has a structure like:

```ts
new FormArray([
  new FormGroup({
    firstName: new FormControl(),
    lastName: new FormControl(),
  }),
  // may repeat ...
])
```

Is this correct?

Why not simply pass a `FormArray` to your custom component as an input? Or, alternatively, if the `FormArray` in question is part of a `FormGroup`, you could specify the control name as an input on the custom component, inject the `ControlContainer` into the custom component, and have the component grab the `FormArray` from the `FormGroup` using the specified control name. Etc.

Or are you simply saying that you don't know how to accomplish your goal?

**Comment by junaidbinsarfraz** [comment]
You are right, I am looking for a way to pass `FormArray` from parent component and that `FormArray` will have its own structure (like `FormControl`s and/or `FormArray`s). So it will be like 

**Parent**
_TS_
```
this.formBuilder.group({
  'field1': [],
  'field2': [],
  'child': this.formBuilder.array([ // Its own FormControls and/or FormArray passed to child ])
});
```
_HTML_
```
<parent>
  ... other tags
  <app-child formControlName="child"></app-child>
</parent>
```

**Child**
_TS_
```
export class ChildComponent implements ControlValueAccessor {
 
  // Other functions and variables
  
  myFunc() {
    // Somewhere in the class function
    this.propagateChange(this.myFormArray);
  }

  propagateChange = (_: any) => {};

  writeValue(obj: any): void {
    // initialized formArray
  }

  registerOnChange(fn: any): void {
    this.propagateChange = fn;
  }

  registerOnTouched(fn: any): void {  }

  setDisabledState?(isDisabled: boolean): void {  }
  
}
```

**Issue**
I have tried to send FormArray to child component using multiple ways but none of them succeeded.

1. TS: 
`'child': this.formBuilder.array([])` 
then pass it using HTML 
`<app-child formControlName="child"></app-child>` 
-- **The issue** with this approach is that it throws error `ERROR TypeError: "control.registerOnChange is not a function"`
2. TS: 
`'child': [this.formBuilder.array([])] // as FormControl` 
then pass it using HTML
`<app-child formControlName="child"></app-child>` 
-- **The issue** with that it is not treated as proper `FormArray`, means it is `FormControl `not `FormArray`. This approach is absolutely not right way to do it. Also validations not work properly.
3. TS: 
`'child': this.formBuilder.array([])` 
then pass it using HTML 
`<app-child formArrayName="child"></app-child>` 
-- **The issue** with this approach is that `writeValue` and `propagateChange` functions do not work/call because `FormArray` is sent as `formArrayName `rather then `formControlName`.  So as soon as I submit the form, the `FormArray `is always empty. Not correct approach I guess.

After these failures I have tried few other lame approaches (which I don't recall right now) -- but they also failed.
I think `ControlValueAccessor `does not support `FormArray`, or I couldn't figure it out.

**Comment by thefliik** [comment]
Gotcha, ya `ControlValueAccessor` doesn't really have anything to do with what you're trying to accomplish.

While there are a number of ways to accomplish your goal, the easiest way is probably to simply pass your FormArray to the child component as a regular `@Input()` property. For Example:

```ts
@Component({
  selector: `app-child`,
  template: `
    <h1>Child</h1>
    <div *ngFor="let nameGroup of control.controls" >
      <input [formControl]="nameGroup.get('firstName')">
      <input [formControl]="nameGroup.get('lastName')">
    </div>
  `
})
export class ChildComponent {
  @Input() control: FormArray;
}
```

Usage like:

```html
<parent>
  ... other tags
  <app-child [control]="theFormArray"></app-child>
</parent>
```

**Comment by junaidbinsarfraz** [comment]
I thought of that way too, but my requirements are quite complex. Above example that I listed was merely for understanding. Let me explain you what exactly I am looking for.

**Parent**
_TS_
```
// Top Form
this.formBuilder.group({
  'f1', [],
  'anArray': this.formBuilder.array([
    this.formBuilder.group({ 'field1': [], 'field2': [], child: this.formBuilder.array([// Add few child array]),
    this.formBuilder.group({ 'field1': [], 'field2': [], child: this.formBuilder.array([// Add few child array]),
    ....
    ])
})
```
_HTML_
```
<div>
  <form [formGroup]="topForm">
    <input formControlName="f1" />
    <div formArrayName="anArray" class="mt-3" *ngFor="let value of anArray.controls; let i = index;">
      <div [formGroupName]="i" >
        <input formControlName="field1" />
        <input formControlName="field2" />
        <app-child formControlName="child"></app-child> <!-- here child will be an array -->
      </div>
    </div>
  </form>
</div>
```
Now here is the scenario
1. There can be `n` numbers of `anArray` which can be dynamically added. 
2. According to your suggestion, I have to create extra list/Map of `FormArray `for all  `ChildComponent `that are linked to respective `anArray`. Then how can I reflect/attach that in HTML for respective `anArray`?
3. Moreover your suggestion will be (not impossible) but complex to maintain. Like I have to handle lots of logic in parent component (which should be not needed manually if we have something like `ArrayValueAccessor`). 
4. Also managing validations on `ChildComponent `linked to `ParentComponent `will be quite difficult (like `submitButton `of whole form should be disabled until all `ChildComponents `are not filled/validated).

That's why I suggest Angular team to either include `FormArray `in `ControlValueAccessor `or make a new interface for `FormArray`.

**Comment by thefliik** [comment]
This example, combined with the code from my previous comment, does what you want it to.

```html
<div>
  <form [formGroup]="topForm">
    <input formControlName="f1" />
    <div class="mt-3" *ngFor="let control of topForm.get('anArray')!.controls">
      <input [formControl]="control.get('field1')" />
      <input [formControl]="control.get('field2')" />
      <app-child [control]="control.get('child')"></app-child>
    </div>
  </form>
</div>
```

**Comment by junaidbinsarfraz** [comment]
Seems promising! 
I will give it a try. If any issue faced, I will get back to you. 
Thanks

**Comment by maxime1992** [comment]
@junaidbinsarfraz if your goal is to easily build sub form, you can take a look into `ngx-sub-form`: https://github.com/cloudnc/ngx-sub-form

Behind the scenes it does use `ControlValueAccessor` but it handles all the boilerplate (+ some angular know issues...) internally.

Also, what you're trying to achieve with FormArray is totally doable and here's an example from the doc: https://github.com/cloudnc/ngx-sub-form#dealing-with-arrays

**Comment by lazarljubenovic** [comment]
If your child component has an array of controls, then it's not a control. It would be an anti-pattern to have form controls inside something that's a form control itself: this is what form groups (`FormGroup`) and form arrays (`FormArray`) have been created for. Your component is simply a component that should somehow be made aware that its controls are part of an externally created form. Not a control itself.

I haven't checked out `ngx-sub-form` which @maxime1992 suggested above, but I guess it's a nice wrapper around a technique I've described in this StackOverflow answer: https://stackoverflow.com/a/46025197/2131286.

Basically, if you put a `formGroup`, `formGroupName`, `formArray` or `formArrayName` on a component, then that component can pick up something called Control Container through injection. Inject `ControlContainer` in the reusabile component, and `ControlContainer#control` will give you the control, decided by `formArrayName` or similar.
